compiler/shared_ast/typing.ml

```diff
@@ -373,56 +396,95 @@ and typecheck_expr_top_down :
     Expr.elocation loc (uf_mark (ast_to_typ ty))
   | A.EStruct { name; fields } ->
     let mark = ty_mark (TStruct name) in
-    let str = A.StructMap.find name ctx.A.ctx_structs in
+    let str = A.StructName.Map.find name ctx.A.ctx_structs in
     let _check_fields : unit =
       let missing_fields, extra_fields =
-        A.StructFieldMap.fold
+        A.StructField.Map.fold
           (fun fld x (remaining, extra) ->
-            if A.StructFieldMap.mem fld remaining then
-              A.StructFieldMap.remove fld remaining, extra
-            else remaining, A.StructFieldMap.add fld x extra)
+            if A.StructField.Map.mem fld remaining then
+              A.StructField.Map.remove fld remaining, extra
+            else remaining, A.StructField.Map.add fld x extra)
           fields
-          (str, A.StructFieldMap.empty)
+          (str, A.StructField.Map.empty)
       in
       let errs =
         List.map
           (fun (f, ty) ->
-            ( Some
-                (Format.asprintf "Missing field %a" A.StructFieldName.format_t f),
+            ( Some (Format.asprintf "Missing field %a" A.StructField.format_t f),
               Marked.get_mark ty ))
-          (A.StructFieldMap.bindings missing_fields)
+          (A.StructField.Map.bindings missing_fields)
         @ List.map
             (fun (f, ef) ->
-              let dup = A.StructFieldMap.mem f str in
+              let dup = A.StructField.Map.mem f str in
               ( Some
                   (Format.asprintf "%s field %a"
                      (if dup then "Duplicate" else "Unknown")
-                     A.StructFieldName.format_t f),
+                     A.StructField.format_t f),
                 Expr.pos ef ))
-            (A.StructFieldMap.bindings extra_fields)
+            (A.StructField.Map.bindings extra_fields)
       in
       if errs <> [] then
         Errors.raise_multispanned_error errs
           "Mismatching field definitions for structure %a" A.StructName.format_t
           name
     in
     let fields' =
-      A.StructFieldMap.mapi
+      A.StructField.Map.mapi
         (fun f_name f_e ->
-          let f_ty = A.StructFieldMap.find f_name str in
+          let f_ty = A.StructField.Map.find f_name str in
           typecheck_expr_top_down ctx env (ast_to_typ f_ty) f_e)
         fields
     in
     Expr.estruct name fields' mark
+  | A.EDStructAccess { e = e_struct; name_opt; field } ->
+    let t_struct =
+      match name_opt with
+      | Some name -> TStruct name
+      | None -> TAny (Any.fresh ())
+    in
+    let e_struct' =
+      typecheck_expr_top_down ctx env (unionfind t_struct) e_struct
+    in
+    let name =
+      match UnionFind.get (ty e_struct') with
+      | TStruct name, _ -> name
+      | TAny _, _ ->
+        Printf.ksprintf failwith
+          "Disambiguation failed before reaching field %s" field
```

Comment:
OK so you put the code for disambiguation here. I would have preferred to see it in [compiler/desugared/disambiguate.ml](https://github.com/CatalaLang/catala/pull/361/files/4ee4a96ac7e9f12fe5cf1e7a4be8b78ad624d472#diff-3920d1db78c17509179cc641fe562122da0172f6bcd9ae0b2c827dfa65470f33). Do you think you can switch it?

Reply:
The two are pretty inter-mingled: the thing is that for typing an expression like `struct.field1.field2`, you will actually need to disambiguate and resolve `field1` to get the structure type for `field2`.

That's why my first draft, which would only locally call the typer when a disambiguation was required, would end up duplicating most of the logic and re-typing the same sub expressions multiple times.

Another approach could be to expose much more of `Typing` and find a way for external disambiguation code to be called from within the typer, but I fear the complexity of that would make things worse overall.

Reply:
yeah after thinking more about it I agree with you